docs/testing-guidelines.md

```diff
@@ -0,0 +1,121 @@
+# Testing Guidelines for UDS Marketplace Apps
+
+## Table of Contents
+
+1. [Introduction](#introduction)
+2. [Journey Testing](#journey-testing)
+   - [Definition](#definition)
+   - [Key Principles](#key-principles)
+   - [Implementation Guidelines](#implementation-guidelines)
+   - [Tools](#tools)
+3. [Upgrade Testing](#upgrade-testing)
+   - [Key Considerations](#key-considerations)
+4. [Linting and Static Analysis](#linting-and-static-analysis)
+   - [Recommended Tools](#recommended-tools)
+5. [Best Practices](#best-practices)
+6. [Best Practice Examples](#best-practice-examples)
+   - [Journey Tests](#journey-tests)
+   - [Upgrade Tests](#upgrade-tests-1)
+   - [Test Execution](#test-execution)
+7. [Consistency Across Packages](#consistency-across-packages)
+8. [Contribution and Maintenance](#contribution-and-maintenance)
+9. [Related Resources](#related-resources)
+
+## Introduction
+
+This document outlines the testing guidelines and best practices for UDS Packages. These guidelines aim to ensure consistency, reliability, and quality across all of our packages.
+
+## Journey Testing
+
+Journey testing is crucial for ensuring that UDS packages function correctly within their deployment environment and integrate properly with other UDS components.
+
+### **Definition**
+A **Journey** in the context of UDS package testing is defined as:
+> The minimal and least fragile way to test all features of an application that could be broken from how we deploy it.
+
+### **Key Principles**
+- Focus on testing the deployment, not the application itself (that's the upstream developer's job).
+- Ensure that our deployment method doesn't break key functionalities.
+- Test integration with UDS core components (e.g., Istio, Keycloak).
+
+### **Implementation Guidelines**
+- Design tests that cover essential functionality without being overly complex.
+- Focus on deployment-related issues (e.g., network policies, SSO access, cluster resource access).
+- Handle limitations due to licensing or other constraints by documenting them and implementing what testing is possible.
+
+### **Tools**
```

Comment:
I think we should...
1. norm around a single testing tool, playwright seems to be the more popular choice
2. Replace or remove jest. I'm not sure of any instances where we've used Jest in uds-packages (I could very well be wrong)
3. We should lean into Zarf component actions and mention that here.

Reply:
Thanks for your feedback. Here are my thoughts:

1. Standardizing on a single tool: While beneficial, different projects may have varying testing needs. Jest excels at unit and integration testing, particularly in Test-Driven Development (TDD) scenarios, whereas Playwright is specifically designed for end-to-end testing across multiple browsers. I am new to both of these tools and TypeScript, so I could be a bit off-base there.

2. Jest usage: There are a few repos using Jest already. For example [redpanda](https://github.com/defenseunicorns/uds-package-redpanda/blob/main/tests/journey/redpanda.test.ts) and [argocd](https://github.com/defenseunicorns/uds-package-argocd/blob/main/tests/journey/argocd.api.test.ts) I'm open to recommending one over the other, but we should clearly define when to use each or if there's a use case for both, to keep it consistent. I also think if we decide that only one should be used that we ensure repos adhere to it.

3. Defining tool usage: Agreed. We should add a "When to use Jest vs Playwright" section to guide teams if that is what we align on.

4. Zarf component actions: Good point. I'll incorporate this into our guidelines. Is there any "shining example" I can reference?

Proposal:
a) Keep both Jest and Playwright
b) Add a section on when to use each
c) Include Zarf component actions info

I think this would keep it flexible while also providing a bit clearer direction. Thoughts?

Reply:
Your proposal sounds great!